src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+                var allArguments = invocationArguments.Insert(insertionIndex, newMethodInvocation);
+                return (TInvocationExpressionSyntax)editor.Generator.InvocationExpression(expressionFromInvocation, allArguments);
+            }
+        }
+
+        /// <summary>
+        /// Goes through all of the invocations and replaces the expression with identifiers from the invocation 
+        /// arguments and rewrites the call site with the updated expression as a new argument.
+        /// </summary>
+        private static async Task<Solution> RewriteSolutionWithNewMethodAsync(
+            Document originalDocument, TExpressionSyntax expression, IMethodSymbol methodSymbol,
+            SyntaxNode containingMethod, bool allOccurrences, string parameterName,
+            ImmutableDictionary<Document, ImmutableArray<TInvocationExpressionSyntax>> callSites, bool trampoline,
+            CancellationToken cancellationToken)
+        {
+            var modifiedSolution = originalDocument.Project.Solution;
+
+            foreach (var (project, projectCallSites) in callSites.GroupBy(kvp => kvp.Key.Project))
+            {
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var (document, invocationExpressionList) in projectCallSites)
+                {
+                    modifiedSolution = await ModifyDocumentInvocationsTrampolineAndIntroduceParameterAsync(compilation,
+                        modifiedSolution, document, originalDocument, invocationExpressionList, methodSymbol, containingMethod,
+                        allOccurrences, parameterName, expression, trampoline, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            return modifiedSolution;
+        }
+
+        /// <summary>
+        /// This method iterates through the arguments at the invocation location and maps them back
+        /// to the identifiers in the expression to create a new expression as an argument.
+        /// </summary>
+        private (TExpressionSyntax, bool) CreateNewArgumentExpression(SemanticModel invocationSemanticModel,
+            Document document, ISyntaxFactsService syntaxFacts, Dictionary<TIdentifierNameSyntax, IParameterSymbol> mappingDictionary,
+            SyntaxGenerator generator, TExpressionSyntax expression, TIdentifierNameSyntax variable,
+            SeparatedSyntaxList<SyntaxNode> invocationArguments, SeparatedSyntaxList<SyntaxNode> updatedInvocationArguments,
+            bool parameterIsNamed, CancellationToken cancellationToken)
+        {
+            if (mappingDictionary.TryGetValue(variable, out var mappedParameter))
+            {
+                var parameterMapped = false;
+                var oldNode = expression.GetCurrentNode(variable);
+                RoslynDebug.AssertNotNull(oldNode);
+                for (var i = 0; i < invocationArguments.ToArray().Length; i++)
```

Comment:
this is bad.  on each loop you are doing a ToArray, just to get the length of the arguments.  just literally call invocationArguments.COunt.

Reply:
yea that was poor coding, removed the loop entirely anyways